### PR TITLE
Kernel/Interrupts: Initialize two spurious handlers when PIC is disabled

### LINKED ISSUE #12134: Kernel: Bare-metal PICs can generate spurious interrupts even when fully masked
Some bare-metal motherboards generate spurious interrupts through the PIC even when all PIC interrupts are masked. A possible fix is installing a couple of handlers that simply ignore these interrupts in the unused interrupt space. (32 - 80 on serenity currently)

cc @supercomputer7

### PR DESCRIPTION
**Pull request by supercomputer7**
Even if PIC was disabled it can still generate noise (spurious IRQs) so
we need two handlers being registered for handling such case.

Fixes #12134.

### CODE REVIEW

File: Kernel/Interrupts/InterruptManagement.cpp [hunk 1]
```diff
@@ -98,6 +98,19 @@ u8 InterruptManagement::get_irq_vector(u8 mapped_interrupt_vector)
     return mapped_interrupt_vector;
 }
 
+RefPtr<IRQController> InterruptManagement::get_responsible_irq_controller(IRQControllerType controller_type, u8 interrupt_vector)
+{
+    if (m_interrupt_controllers.size() == 1 && m_interrupt_controllers[0]->type() == IRQControllerType::i8259) {
+        return m_interrupt_controllers[0];
+    }
```
Comment (IdanHo) [inline Kernel/Interrupts/InterruptManagement.cpp:105]: No need for the curly braces, but I'm not entirely sure what the point of this if is? Is this an optimization?

File: Kernel/Interrupts/SpuriousInterruptHandler.h [hunk 2]
```diff
@@ -32,11 +34,14 @@ class SpuriousInterruptHandler final : public GenericInterruptHandler {
...
     virtual StringView controller() const override;
 
+    void enable_interrupt_vector_for_disabled_pic();
+
 private:
     void enable_interrupt_vector();
     void disable_interrupt_vector();
     explicit SpuriousInterruptHandler(u8 interrupt_number);
     bool m_enabled { false };
+    bool m_enabled_for_disabled_pic { false };
```
Comment (IdanHo) [inline Kernel/Interrupts/SpuriousInterruptHandler.h:44]: This boolean is not actually used for anything AFAICT

File: Kernel/Interrupts/SpuriousInterruptHandler.cpp [hunk 3]
```diff
@@ -16,6 +17,22 @@ UNMAP_AFTER_INIT void SpuriousInterruptHandler::initialize(u8 interrupt_number)
     handler->register_interrupt_handler();
 }
 
+void SpuriousInterruptHandler::initialize_for_disabled_master_pic()
+{
+    auto* handler = new SpuriousInterruptHandler(7);
+    // Note: When we disable the PIC, the IRQ base offset is set to 0x20
+    register_generic_interrupt_handler(0x20 + 7, *handler);
```
Comment (IdanHo) [inline Kernel/Interrupts/SpuriousInterruptHandler.cpp:24]: Can we make these comments useless by making `0x20` a constexpr constant and using it both here and in the PIC code?

File: Kernel/Interrupts/InterruptManagement.cpp [hunk 4]
```diff
@@ -98,6 +98,19 @@ u8 InterruptManagement::get_irq_vector(u8 mapped_interrupt_vector)
...
+    for (auto& irq_controller : m_interrupt_controllers) {
+        if (irq_controller->gsi_base() <= interrupt_vector)
+            if (irq_controller->type() == controller_type)
```
Comment (Panky-codes) [inline Kernel/Interrupts/InterruptManagement.cpp:108]: Couldn't we just combine these two `if` statements with a one `if` statement with &&:
```cpp
        if (irq_controller->gsi_base() <= interrupt_vector && irq_controller->type() == controller_type)
```

File: Kernel/Interrupts/SpuriousInterruptHandler.cpp (hunk 3 above)
Comment (supercomputer7) [reply Kernel/Interrupts/SpuriousInterruptHandler.cpp:24]: I rather do that in a different pull request so I don't need to touch that code for now.

Comment (IdanHo) [reply Kernel/Interrupts/SpuriousInterruptHandler.cpp:24]: I don't think a different pull request is appropriate. That constant was only used in one location, which makes hardcoding it ok, but now it's used in multiple locations, so repeating it is an issue. You can create a separate commit that extract the constant into a constexpr/define and then use it in your current commit.

Comment (supercomputer7) [reply Kernel/Interrupts/SpuriousInterruptHandler.cpp]: I used a constexpr, hopefully it's good enough :)

File: Kernel/Interrupts/PIC.h [hunk 5]
```diff
@@ -10,6 +10,9 @@
 #include <Kernel/Interrupts/IRQController.h>
 
 namespace Kernel {
+
+static constexpr size_t g_pic_disabled_vector_base = 0x20;
```
Comment (IdanHo) [inline Kernel/Interrupts/PIC.h:14]: We usually don't prefix constexpr constants with `g_`

File: Kernel/Interrupts/PIC.cpp [hunk 6]
```diff
@@ -148,7 +148,7 @@ void PIC::complete_eoi() const
 void PIC::hard_disable()
```
Comment (IdanHo) [inline Kernel/Interrupts/PIC.cpp:148]: commit title nitpick: `a disabled` => `the disabled`

File: Kernel/Arch/x86/common/Interrupts.cpp [hunk 7]
```diff
@@ -640,6 +640,54 @@ UNMAP_AFTER_INIT void idt_init()
         register_interrupt_handler(i, unimp_trap);
 
     dbgln("Initializing unhandled interrupt handlers");
+    register_interrupt_handler(0x20, interrupt_32_asm_entry);
+    register_interrupt_handler(0x21, interrupt_33_asm_entry);
```
Comment (IdanHo) [inline Kernel/Arch/x86/common/Interrupts.cpp:644]: If we're making these valid interrupts, the for loop above should be removed.

Comment (supercomputer7) [reply Kernel/Arch/x86/common/Interrupts.cpp:644]: Not really, we still need this from 0x11 to 0x1f. I will fix this now.

File: Kernel/Interrupts/PIC.h [hunk 8]
```diff
@@ -10,6 +10,12 @@
 #include <Kernel/Interrupts/IRQController.h>
 
 namespace Kernel {
+
+static constexpr size_t pic_disabled_vector_base = 0x90;
+static constexpr size_t pic_disabled_vector_end = 0x9f;
+static_assert(pic_disabled_vector_base < GENERIC_INTERRUPT_HANDLERS_COUNT);
+static constexpr size_t pic_enabled_vector_base = 0x50;
```
Comment (IdanHo) [inline Kernel/Interrupts/PIC.h:17]: What's the difference between IRQ_VECTOR_BASE and pic_enabled_vector_base?

File: Kernel/Interrupts/PIC.h [hunk 9]
```diff
@@ -10,6 +10,12 @@
 #include <Kernel/Interrupts/IRQController.h>
 
 namespace Kernel {
+
+static constexpr size_t pic_disabled_vector_base = 0x90;
+static constexpr size_t pic_disabled_vector_end = 0x9f;
+static_assert(pic_disabled_vector_base < GENERIC_INTERRUPT_HANDLERS_COUNT);
```
Comment (IdanHo) [inline Kernel/Interrupts/PIC.h:16]: What is the meaning of this static assert?

File: Kernel/Arch/x86/common/Interrupts.cpp [hunk 10]
```diff
@@ -496,6 +498,23 @@ void handle_interrupt(TrapFrame* trap)
...
+        } else if (irq == 15) {
+            handler = s_disabled_interrupt_handler[1];
+        }
+        VERIFY(handler);
+        handler->increment_invoking_counter();
+        handler->handle_interrupt(regs);
+        handler->eoi();
+        return;
+    }
     u8 irq = (u8)(regs.isr_number - 0x50);
```
Comment (IdanHo) [inline Kernel/Arch/x86/common/Interrupts.cpp:518]: This 0x50 should be `IRQ_VECTOR_BASE` no?

File: Kernel/Arch/x86/common/Interrupts.cpp [hunk 11]
```diff
@@ -496,6 +498,23 @@ void handle_interrupt(TrapFrame* trap)
...
+            handler = s_disabled_interrupt_handler[0];
+        } else if (irq == 15) {
+            handler = s_disabled_interrupt_handler[1];
+        }
+        VERIFY(handler);
+        handler->increment_invoking_counter();
+        handler->handle_interrupt(regs);
+        handler->eoi();
+        return;
+    }
```
Comment (IdanHo) [inline Kernel/Arch/x86/common/Interrupts.cpp:517]: I don't leave this (both the hardcoded 7/15 and the duplicated increment_invoking_counter/handle_interrupt/eoi, but as you said you intend to refactor this, I guess it's fine.

Comment (supercomputer7) [reply Kernel/Arch/x86/common/Interrupts.cpp:517]: I will keep a note for this when I refactor the low-level interrupt handling code.

Comment (supercomputer7) [reply Kernel/Arch/x86/common/Interrupts.cpp]: Fixed this now, I think it's fine until we do a refactor for this code :)

File: Kernel/Arch/x86/ISRStubs.h [hunk 12]
```diff
@@ -11,6 +11,54 @@
 #include <AK/Platform.h>
```
Comment (IdanHo) [inline Kernel/Arch/x86/ISRStubs.h:11]: Commit message nitpicks
`Even if PIC was disabled` => `Even if the PIC was disabled`
`so we need two handlers being registered for handling such case.` => `so we need to register two handlers for handling such cases.`
`we declare interrupt service routine offset 0x90 to 0x9f as` this is not correct anymore
`to a different interrupt handlers at` => `to different interrupt handlers at`
`at different location.` => `at different location.s`

**Pull request merged**